Meziantou.Polyfill.Editor/M;System.Net.Sockets.UdpClient.SendAsync(System.ReadOnlyMemory{System.Byte},System.Net.IPEndPoint,System.Threading.CancellationToken).cs

```diff
@@ -0,0 +1,12 @@
+﻿using System;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+
+static partial class PolyfillExtensions
+{
+	public async static ValueTask<int> SendAsync(this System.Net.Sockets.UdpClient client, ReadOnlyMemory<byte> datagram, IPEndPoint? endPoint, CancellationToken cancellationToken = default)
+	{
+		return await client.SendAsync(datagram.ToArray(), datagram.Length, endPoint).WaitAsync(cancellationToken);
```

Comment:
You should new ValueTask<int>(task) instead of `WaitAsync`.

Reply:
Something like that ? 
```cs
public static ValueTask<int> SendAsync(this System.Net.Sockets.UdpClient client, ReadOnlyMemory<byte> datagram, IPEndPoint? endPoint, CancellationToken cancellationToken = default)
{
    return new(client.SendAsync(datagram.ToArray(), datagram.Length, endPoint));
}
```

What should I do with the cancellationToken ? I do not known.

Reply:
> Something like that ?

Yes

> What should I do with the cancellationToken ? I do not known.

If it's not supported by `client.SendAsync`, then don't use it. The assumption is that the CancellationToken should cancel the network request. If you use it in `WaitAsync`, you are not cancelling the request, you just stop waiting for it. In this case, I prefer to let the caller do it itself.

You can still use it at the beggining of the method

````
if(cancellationToken.IsCancellationRequested)
    return ValueTask.FromCanceled(cancellationToken);
````

Reply:
I did not added 
```cs
        if (cancellationToken.IsCancellationRequested)
            return ValueTask.FromCanceled<int>(cancellationToken);
```
because it is not supported for .Net 4.8.1 (it fail the tests)

Reply:
I can't see how canceling a UDP send would even be a thing. It's pretty much a do or don't operation, and the code literally just ends up calling a Socket.SendAsync, never even touching the CancellationToken along the way.

It's weird that UdpClient.SendAsync even has the parameter in the first place.

Reply:
In fact, the ValueTask<int> overload of UdpClient.SendAsync, specifically, actually makes it optional anyway.

Weird design decisions. I wonder what was ultimately behind that - maybe consistency with other APIs?

Here's the .net 8.0.6 implementation:

https://github.com/dotnet/runtime/blob/v8.0.6/src/libraries/System.Net.Sockets/src/System/Net/Sockets/UDPClient.cs#L563

Reply:
As it turns out, nothing in UdpClient uses a cancellationToken passed to it.

There is no occurrence of `cancellationToken.` in the whole file.